From 1f08023e69c6a9e31ae4248950c4c3389c51f171 Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Mon, 26 Aug 2024 21:41:55 +0800
Subject: [PATCH] WLANRTSYS-84268 Avoid pruning of own STA entry if Single Link
 ML STA

Issue:
When a previously associated ML STA tries to connect again as
Single Link STA with same MLD and Link Address, pruning of its
own entry occured by fetching the linked STA of the current Single
link STA in sibling hapd.

Fix:
Do MLD cleanup of the linked STA only if its STA entry is not
equal to the current STA entry.

FIXUP to WLANRTSYS-62845 AP MLD Auth Assoc and eapol changes
---
 src/ap/mxl_utils.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/ap/mxl_utils.c b/src/ap/mxl_utils.c
index 8db9d58d8..e8daf76f2 100644
--- a/src/ap/mxl_utils.c
+++ b/src/ap/mxl_utils.c
@@ -110,7 +110,7 @@ void mxl_mlo_hostapd_prune_associations_mld_mac(struct hostapd_data *hapd, struc
 			mxl_ml_stop_sta_traffic_remove_mld(hapd, prev);
 			linked_sta = prev->mxl_sta_info.linked_sta;
 			mxl_mlo_cleanup_mld_sta(hapd, prev);
-			if (linked_sta)
+			if (linked_sta && (linked_sta != new_sta))
 				mxl_mlo_cleanup_mld_sta(hapd->mxl_data.sibling_hapd, linked_sta);
 		}
 	}
-- 
2.43.0

